server/src-lib/Hasura/GraphQL/Schema.hs

```diff
@@ -159,32 +166,70 @@ buildGQLContext =
+          ucQueries   <- finalizeParser <$> unauthenticatedQueryWithIntrospection qRemotes mRemotes
+          ucMutations <- fmap finalizeParser <$> unauthenticatedMutation mRemotes
           pure $ GQLContext ucQueries ucMutations
 
-        -- | The 'query' type of the remotes. TODO: also expose mutation
-        -- remotes. NOT TODO: subscriptions, as we do not yet aim to support
+        buildRoleBasedRemoteSchemaParser :: RoleName -> RemoteSchemaCache -> m [ParsedIntrospection]
+        buildRoleBasedRemoteSchemaParser role remoteSchemaCache = do
+          let remoteSchemaIntroInfos = map fst $ toList remoteSchemaCache
+          remoteSchemaPerms <-
+            for remoteSchemaIntroInfos $ \(RemoteSchemaCtx _ ctx perms) ->
+              case Map.lookup role perms of
+                Nothing -> return Nothing
+                Just introspectRes -> do
+                  (queryParsers, mutationParsers, subscriptionParsers) <-
+                     P.runSchemaT @m @(P.ParseT Identity) $ buildRemoteParser introspectRes $ rscInfo ctx
+                  return $ Just $ ParsedIntrospection queryParsers mutationParsers subscriptionParsers
+          return $ catMaybes remoteSchemaPerms
+
+        -- | The 'query' type of the remotes.
+        --  NOT TODO: subscriptions, as we do not yet aim to support
         -- these.
-        queryRemotes = concatMap (piQuery . snd) remotes
-        mutationRemotes = concatMap (concat . piMutation . snd) remotes
-        queryHasuraOrRelay = case queryType of
+        adminQueryRemotes = concatMap (piQuery . snd) remotes
+        -- | The 'mutation' type of the remotes.
+        adminMutationRemotes = concatMap (concat . piMutation . snd) remotes
+
+        queryRemotes
+          :: [ParsedIntrospection]
+          -> [P.FieldParser (P.ParseT Identity) RemoteField]
+        queryRemotes = concatMap piQuery
+
+        mutationRemotes
+          :: [ParsedIntrospection]
+          -> [P.FieldParser (P.ParseT Identity) RemoteField]
+        mutationRemotes = concatMap (concat . piMutation)
+
+        queryHasuraOrRelay (qRemotes, mRemotes) = case queryType of
           QueryHasura -> queryWithIntrospection (Set.fromMap $ graphQLTables $> ())
-                         validFunctions queryRemotes mutationRemotes
+                         validFunctions qRemotes mRemotes
                          allActionInfos nonObjectCustomTypes
           QueryRelay  -> relayWithIntrospection (Set.fromMap $ graphQLTables $> ()) validFunctions
 
         buildContextForRoleAndScenario :: RoleName -> Scenario -> m GQLContext
         buildContextForRoleAndScenario roleName scenario = do
           SQLGenCtx{ stringifyNum } <- askSQLGenCtx
+          roleBasedRemoteSchemas <-
+             if | (roleName == adminRoleName) || (not isEnabledRemoteSchemaPerms) -> pure $ map snd remotes
+                | otherwise -> buildRoleBasedRemoteSchemaParser roleName allRemoteSchemas
```

Comment:
Maybe, instead of `otherwise`, we can use `isEnabledRemoteSchemaPerms` to be concise.